lib/messages.js

```diff
@@ -1,303 +1,107 @@
-// This file is auto generated by the protocol-buffers compiler
+const cenc = require('compact-encoding')
 
-/* eslint-disable quotes */
-/* eslint-disable indent */
-/* eslint-disable no-redeclare */
-/* eslint-disable camelcase */
-
-// Remember to `npm install --save protocol-buffers-encodings`
-var encodings = require('protocol-buffers-encodings')
-var varint = encodings.varint
-var skip = encodings.skip
-
-exports.TYPE = {
-  QUERY: 1,
-  UPDATE: 2,
-  RESPONSE: 3
+const IPv4 = exports.IPv4 = {
+  preencode (state, ip) {
+    state.end += 4
+  },
+  encode (state, ip) { // TODO: move over fast parser from ./id.js
+    const nums = ip.split('.')
+    state.buffer[state.start++] = Number(nums[0]) || 0
+    state.buffer[state.start++] = Number(nums[1]) || 0
+    state.buffer[state.start++] = Number(nums[2]) || 0
+    state.buffer[state.start++] = Number(nums[3]) || 0
+  },
+  decode (state) {
+    if (state.end - state.start < 4) throw new Error('Out of bounds')
+    return state.buffer[state.start++] + '.' + state.buffer[state.start++] + '.' + state.buffer[state.start++] + '.' + state.buffer[state.start++]
+  }
 }
 
-var Holepunch = exports.Holepunch = {
-  buffer: true,
-  encodingLength: null,
-  encode: null,
-  decode: null
+const peerIPv4 = exports.peerIPv4 = {
+  preencode (state, peer) {
+    state.end += 6
+  },
+  encode (state, peer) {
+    IPv4.encode(state, peer.host)
+    cenc.uint16.encode(state, peer.port)
+  },
+  decode (state) {
+    return {
+      host: IPv4.decode(state),
+      port: cenc.uint16.decode(state)
+    }
+  }
 }
 
-var Message = exports.Message = {
-  buffer: true,
-  encodingLength: null,
-  encode: null,
-  decode: null
-}
+const peerIPv4Array = exports.peerIPv4Array = cenc.array(peerIPv4)
 
-defineHolepunch()
-defineMessage()
+const IS_REQUEST = 0b0001
+const HAS_ID = 0b0010
+const HAS_TOKEN = 0b0100
+const ROUTE_INFO = 0b1000 | IS_REQUEST
+const HAS_TARGET = ROUTE_INFO | IS_REQUEST
+const HAS_CLOSER_NODES = ROUTE_INFO ^ IS_REQUEST
 
-function defineHolepunch () {
-  Holepunch.encodingLength = encodingLength
-  Holepunch.encode = encode
-  Holepunch.decode = decode
+exports.message = {
+  preencode (state, m) {
```

Comment:
preen code or pre encode?

Reply:
nm it doesn't matter, it's internal, we know what it is

Reply:
"pre encode" but this is part of an api that uses it as one work "preencode"